### DISCUSSION TITLE
Disable autopair of ' in a comment

### DISCUSSION THREAD

**Discussion by eric-hemasystems** [orbitalquark/textadept#266] [Show and tell]
**UPDATE: Code below now also disabled auto-pairs for text-like documents such as markdown or plain text in addition to still disabling it in comments.**

----

Love the auto-pair feature but a small pet peeve of mine was when a `'` was used in a comment it was most often used for a contraction instead of quoting. This means after hitting `'` I almost always had to then delete the unnecessary extra `'`. Would rather for that specific character it not auto-pair in that context.

Hacked together a bit of code that seems to have accomplished that. Sharing here in case it's useful to others. Also I'm a lua and textadept newbie so any feedback is always appreciated:

```lua
-- Default to only doing this in comments but always do it in certian file formats
local always = false
events.connect(events.LEXER_LOADED, function(name)
  always_formats = { 'markdown', 'text' }
  always = table.contains(always_formats, function(lang) return name == lang end)
end)

local auto_pairs
events.connect(events.KEYPRESS, function(code)
  -- Return early if not a '
  if code ~= string.byte("'") then return end

  if not always then
    -- Look at prev character since end of comment line not considered in comment
    local style = buffer:name_of_style(buffer.style_at[buffer.current_pos-1])

    -- Return early if not a comment
    if( style ~= 'comment' ) then return end
  end

  -- Disable auto_pairs storing config in closure for later restore
  auto_pairs, textadept.editing.auto_pairs = textadept.editing.auto_pairs, nil
end)

events.connect(events.CHAR_ADDED, function()
  -- Auto-pairs config not stored so must not have been disabled. Return early
  if not auto_pairs then return end

  -- Auto-pairs was disabled for a char, restore config
  textadept.editing.auto_pairs = auto_pairs

  -- Reset closure variable to avoid restore attempt until another disable happens
  auto_pairs = nil
end)
```

At first I tried to just use the `CHAR_ADDED` event. The problem was it is the same event that [Textadept itself uses](https://github.com/orbitalquark/textadept/blob/default/modules/textadept/editing.lua#L140-L150) to implement the feature. This means my event handler had to actually delete the character instead of preventing it from being inserted in the first place which lead to a weird experience. Secondary the `style_at` seemed to be incorrect in the context of `CHAR_ADDED`. I am guessing the lexer is rebuilding that info? If I used [`timeout`](https://orbitalquark.github.io/textadept/api.html#timeoutinterval-f-) that allowed me to delay querying that info until a bit later which worked but contributed to it not feeling quite right even more.

Therefore took a different approach of adding two event handlers. One early when the key is pressed to determine if auto_pair should just be disabled (to prevent the extra `'` in the first place). Another to restore auto_pair if it was disabled. This seems to give it a much more natural experience.

**Comment by rgieseke** [comment]
Thanks for sharing, always good to have examples for these kind of adjustments out there!

I haven't tried this out, but did you see that `events.connect(event, f, index)` has an index parameter (https://orbitalquark.github.io/textadept/api.html). You can run your event handler before Textadept's.

You could stop propagation if you detected that you were in a comment:

> If any handler explicitly returns a value that is not nil, emit() returns that value and ceases to call subsequent handlers. This is useful for stopping the propagation of an event like a keypress after it has been handled,

**Reply by eric-hemasystems** [reply]
Good information!

That could work although I worry it would also stop other behavior dependent on the `CHAR_ADDED` event that I do want to continue. I really only want to target stopping the auto-pair.

I might change my first handler though to be injected at the 1st position on the `CHAR_ADDED` event. Still do the disable of auto_pairs and have a second event that runs at the end to restore it. Basic logic would be the same as above only it would cut down on my event handler running for key presses that don't result in character insert (i.e. shift, backspace, etc).

**Reply by eric-hemasystems** [reply]
Although I wonder if I would then still have the issue of the `style_at` not being valid and therefore perhaps still running on `KEYPRESS` is best even if it runs a bit more often. Will have to test.

**Reply by rgieseke** [reply]
Yes, this can be tricky, but there aren't too many events running on `char_added` by default: https://github.com/orbitalquark/textadept/search?q=char_added

(Other modules you might use could bring other `char_added` though.) 

Your disable-reenable approach could also do the trick.

**Comment by orbitalquark** [comment]
FYI you can manually tell the lexer to update styling information in real-time using:

  buffer:colorize(1, buffer.current_pos)

Then any subsequent calls to `buffer.style_at[]` will be correct.

Regardless, thanks for sharing this neat trick!
